Concatenate CICE daily output

anton-seaice opened the pull request:
Add script to concatenate daily cice output into one file per month (still containing daily data) and delete the individual daily files. This relies on adding nco to the payu environment per https://github.com/ACCESS-NRI/payu-condaenv/pull/24.

Following Aidan's suggestion, the script is taken from https://github.com/COSIMA/1deg_jra55_ryf/blob/master/sync_data.sh#L87-L108, and on that basis I haven't tested beyond checking that it concatenates data.

The only change I made was change the netcdf output type to -4 (netcdf4) instead of -7 (netcdf4_classic).

Looking at ncdump of output looks correct, i.e. it shows time and time bound dimensions of length 31 days for January.

Maybe @aekiss would like to review too?

aidanheerdegen commented on tools/concat_ice_daily.sh:
```diff
@@ -0,0 +1,24 @@
+#script inspired from https://github.com/COSIMA/1deg_jra55_ryf/blob/master/sync_data.sh#L87-L108
+
+for d in archive/output*/ice/OUTPUT; do
+    for f in $d/iceh.????-??-01.nc; do
+        if [[ ! -f ${f/-01.nc/-IN-PROGRESS} ]] && [[ ! -f ${f/-01.nc/-daily.nc} ]];
+        then
+            touch ${f/-01.nc/-IN-PROGRESS}
+            echo "doing ncrcat -O -L 5 -4 ${f/-01.nc/-??.nc} ${f/-01.nc/-daily.nc}"
+            ${PAYU_PATH}/ncrcat -O -L 5 -4 ${f/-01.nc/-??.nc} ${f/-01.nc/-daily.nc} && chmod g+r ${f/-01.nc/-daily.nc} && rm ${f/-01.nc/-IN-PROGRESS}
+            if [[ ! -f ${f/-01.nc/-IN-PROGRESS} ]] && [[ -f ${f/-01.nc/-daily.nc} ]];
+            then
+                for daily in ${f/-01.nc/-??.nc}
+                do
+                    # mv $daily $daily-DELETE  # rename individual daily files - user to delete
+                    rm $daily
+                done
+            else
+                rm ${f/-01.nc/-IN-PROGRESS}
+            fi
+        fi
+    done
```
```suggestion
    # Don't error if there are no matching patterns
    shopt -s nullglob
    
    # Assuming `$d` contains the directory where the data resides
    
    for first_file in $d/iceh.????-??-01.nc
    do
      # Make a list of all files we wish to concatenate
      icefiles=(${first_file/-01.nc/-??.nc})
    
      if [ ${#icefiles[@]} -gt 0 ]
      then
        iceout="${first_file/-*.nc/-daily.nc}"
        ncrcat -O -L 5 -4 "${icefiles[@]}" ${iceout} && rm "${icefiles[@]}"
      fi
    done
```

Personally I prefer to just delete the files if the return status of the `ncrcat` command is ok. Making temporary files ends up introducing extra logic to deal with them.

Note the above is untested, just a suggestion for how to reduce the complexity of the logic.

anton-seaice replied:
> Making temporary files ends up introducing extra logic to deal with them.

This is copied from the COSIMA scripts. I assume the temporary files were needed for some edge case? @aekiss - Do you know why the temporary files were used?

aidanheerdegen replied:
Sorry to badger you @aekiss but I'm also curious if there were cases of data loss that prompted the design you implemented?

anton-seaice replied:
Andrew said it was just for sanity / checking debug in case of failure. I am happy to remove it.

aidanheerdegen reviewed (commented):
It would be good to get an idea of how long it takes to concatenate some of the high res data and post the results in the PR or Issue. 

If it's time-consuming the script could have PBS directives added and run as a `postscript` which is then submitted to the queue

anton-seaice commented:
> It would be good to get an idea of how long it takes to concatenate some of the high res data and post the results in the PR or Issue.

How would one do this? Are there ways to log the PBS "CPU Time" between user scripts?

anton-seaice commented:
> It would be good to get an idea of how long it takes to concatenate some of the high res data and post the results in the PR or Issue.
> 
> If it's time-consuming the script could have PBS directives added and run as a `postscript` which is then submitted to the queue

With one month of 0.1 degree data, this takes ~2.5 minutes to run in the login node. Compared to approx 1.6 hours of Walltime for the model to run. (Amazingly it turns 3.6GB into 1.5GB too!)

This doesn't parallelise, and thats ~2-3% of the walltime, so I guess it is worth worrying about?

(Reducing compression to level 1, reduces the time to ~1m50sec but file size goes up ~6%)

anton-seaice commented on tools/concat_ice_daily.sh:
```diff
@@ -0,0 +1,24 @@
+#concatenate sea-ice daily output
+#script inspired from https://github.com/COSIMA/1deg_jra55_ryf/blob/master/sync_data.sh#L87-L108
+
+for d in archive/output*/ice/OUTPUT; do
```
If we assume that the use might configure the output to be saved in any number of hours, this gets hard ...

Would assume there will always be data saved at 12 hours (i.e. every combination of 1/2/3/4/6/12 hourly would save a 12 hours, then we can find the days by something like `$output_dir/iceh*.????-??-01-43200.nc`). Messy but probably ok.

The other complexity is that CICE timestamps are at the end of the time period. e.g. with hourly data, there is a file with a name at the midnight at the end of the month. So for January, there is a `archive/output001/ice/OUTPUT/iceh_03h.1901-02-01-00000.nc` file made, but this contains January data. 

So I don't know how, in Bash, to make a list of all files for a month working with both those conditions ? We would probable need to use a calendar tool?

aidanheerdegen replied:
Thanks for the thoughtful engagement. It sounds like we should probably shoot for the common use case to begin with and make an issue to [update to a more general form at a later date](https://hitchhikers.fandom.com/wiki/Somebody_Else%27s_Problem_Field).

jo-basevi commented:
Instead of using an `archive` userscript, could instead use the `sync` userscript? This will run prior to running any sync commands, and it has the benefit of running on a job with less resources. Only con I can think of is that it'll only run when `sync` is enabled, but I guess that'll be similar to how the post-script sync-data.sh script originally worked.

aidanheerdegen commented:
> How would one do this? Are there ways to log the PBS "CPU Time" between user scripts?

Ideally find someone running a 0.1 degree experiment that hasn't run Andrew's sync script so the files are all there and "uncollated". Then time how long it takes to run the nco script independently of a `payu` run.

If there is no such experiment, run the model for a few months and generate some output.

But yes, it is possible to time how long a command runs, and how much resources it takes. The shell builtin `time` works ok, but `/usr/bin/time` offers more control over output, especially memory usage IIRC.

aidanheerdegen commented:
> Amazingly it turns 3.6GB into 1.5GB too!

IKR. There is a reason this is worth doing.

> Only con I can think of is that it'll only run when `sync` is enabled, but I guess that'll be similar to how the post-script sync-data.sh script originally worked

That is a good idea thanks @jo-basevi.

I think it does run into the issue that if someone turns `sync` on part way through their experiment not all of the ice data files will be smooshed together (technical term). But we can be explicit that users should set `sync` for this reason (and for other reasons like their data evaporating). We could also give instructions on how to run the command directly on output directories that haven't had their ice files smooshed.

I did wonder if we couldn't define a `collate` option for CICE that did this work. Ultimately it is a nice idea, but not doable in the time frames we have available and we should plough ahead with using a `sync` userscript.

anton-seaice commented:
I think moving to a payu _postscript_ is the best plan, as it runs as a seperate PBS job, this reduces the resources held waiting for a single PE job to complete?

> I did wonder if we couldn't define a `collate` option for CICE that did this work. Ultimately it is a nice idea, but not doable in the time frames we have available and we should plough ahead with using a `sync` userscript.

I think we might get rid of the need for this step in OM3, or at least remove the grid from the CICE output. 

Also - we've added 'nco' as a dependency in some cases. Do we need to document this somewhere (for users who don't use 'vk83' ) ?

anton-seaice commented:
It looks like setting this as a _postscript_ would stop the sync from running?

https://payu.readthedocs.io/en/latest/config.html#postprocessing

anton-seaice commented:
@aidan - I have updated based on the review comments. Back to you. 

I switched to using the system nco module, rahter than adding to payu-env ?

I cleaned up the script to remove the uneeded operations and only check the last archive folder.

jo-basevi commented:
Yeah if postscript is used, and `sync` is enabled, then it won’t rsync the latest output, as payu has no idea on when the `postscript` job completes or whether it modifies the current output. It’ll still rsync up outputs prior to the last one, if they haven’t already been synced.

Also, if syncing is enabled, a `sync` job by default runs on a `copyq` queue. It starts after the main payu run job or collate job, if it’s enabled, has completed. Its resources can be configured similarly to the `collate` config so can be configured to more/less resources if needed.

anton-seaice commented on tools/concat_ice_daily.sh:
```diff
@@ -0,0 +1,18 @@
+#concatenate sea-ice daily output
+#script inspired from https://github.com/COSIMA/1deg_jra55_ryf/blob/master/sync_data.sh#L87-L108
+
+out_dir=$(ls -td archive/output??? | head -1)/ice/OUTPUT #latest output dir only
```
```suggestion
out_dir=$(ls -dr archive/output??? | head -1)/ice/OUTPUT #latest output dir only
```

anton-seaice commented on tools/concat_ice_daily.sh:
```diff
@@ -0,0 +1,18 @@
+for f in $out_dir/iceh.????-??-01.nc; do
+    #concat daily files for this month
+    echo "doing ncrcat -O -L 5 -4 ${f/-01.nc/-??.nc} ${f/-01.nc/-daily.nc}"
+    ncrcat -O -L 5 -4 ${f/-01.nc/-??.nc} ${f/-01.nc/-daily.nc} 
```
```suggestion
    {PAYU_PATH}/ncrcat -O -L 5 -4 ${f/-01.nc/-??.nc} ${f/-01.nc/-daily.nc} 
```

anton-seaice commented on config.yaml:
```diff
@@ -92,7 +92,13 @@ env:
+    
+modules:
+  load:
+    - nco/5.0.5
```
```suggestion
```

aidanheerdegen commented:
Looking down the barrel of adding this to every config, and not being confident we wouldn't have to update it in the future (see conversation about 6 hourly concatenation) I've made a new repo and moved the code to a PR there

https://github.com/ACCESS-NRI/om2-scripts/pull/1

When we've got that merged I'll manually pop it in `vk83` (like I did with `mppnccombine-fast`) and we'll work on a longer term solution later.

Sorry for mucking you about @anton-seaice

anton-seaice commented:
Ok - no worries. Ill put my changes there. Do we still need to update the _config.yaml_ here?

aidanheerdegen commented:
> Do we still need to update the _config.yaml_ here?

Maybe we'll leave this open and just update the `config.yaml` as an exemplar when we have a final location of the concat script.

aidanheerdegen commented:
This was superseded by a script in a separate repo in this PR

https://github.com/ACCESS-NRI/om2-scripts/pull/1
